File: cores/go/internal/mapper/mmc5.go
// mmc5.go - Mapper 5 (MMC5). Port of cores/c/src/mmc5.c.
package mapper

const mmc5PrgBankSize = 8192

func Mmc5Create(m *Mapper, prg, chr []byte, mirroring Mirroring, hasBattery bool) int {
	s := &m.MMC5
	s.PrgSize = uint32(len(prg))
	if s.PrgSize > 0 {
		s.PrgRom = make([]byte, s.PrgSize)
		copy(s.PrgRom, prg)
	}
	if len(chr) == 0 {
		s.ChrSize = 8192
		s.Chr = make([]byte, s.ChrSize)
		s.ChrIsRAM = true
	} else {
		s.ChrSize = uint32(len(chr))
		s.Chr = make([]byte, s.ChrSize)
		copy(s.Chr, chr)
		s.ChrIsRAM = false
	}
	for i := range s.PrgRAM {
		s.PrgRAM[i] = 0
	}
	s.HasBattery = hasBattery
	s.Mirroring = mirroring
	return 0
}

func (s *Mmc5State) prgBankCount() uint32 {
	n := s.PrgSize / mmc5PrgBankSize
	if n == 0 {
		return 1
	}
	return n
}

func mmc5ReadPRG(m *Mapper, addr uint16) uint8 {
	s := &m.MMC5
	if addr >= 0x5C00 && addr < 0x6000 {
		// EXRAM (simplified)
		return 0
	}
	if addr >= 0x6000 && addr < 0x8000 {
		return s.PrgRAM[(addr-0x6000)&(mmc5PrgBankSize-1)]
	}
	if addr < 0x8000 {
		return 0
	}
	local := uint32(addr - 0x8000)
	bankSlot := local / mmc5PrgBankSize
	offset := local % mmc5PrgBankSize
	count := s.prgBankCount()
	switch bankSlot {
	case 0:
		return s.PrgRom[uint32(s.PrgBank1&0x7F)%count*mmc5PrgBankSize+offset]
	case 1:
		return s.PrgRom[uint32(s.PrgBank2&0x7F)%count*mmc5PrgBankSize+offset]
	case 2:
		return s.PrgRom[uint32(s.PrgBank3&0x7F)%count*mmc5PrgBankSize+offset]
	case 3:
		return s.PrgRom[(count-1)*mmc5PrgBankSize+offset]
	}
	return 0
}

func mmc5WritePRG(m *Mapper, addr uint16, value uint8) {
	s := &m.MMC5
	if addr >= 0x5C00 && addr < 0x6000 {
		return
	}
	if addr >= 0x6000 && addr < 0x8000 {
		s.PrgRAM[(addr-0x6000)&(mmc5PrgBankSize-1)] = value
		return
	}
	switch addr {
	case 0x5115:
		s.PrgMode = value & 0x03
	case 0x5120:
		s.ChrBank1 = value
	case 0x5121:
		s.ChrBank2 = value
	case 0x5122:
		s.ChrBank3 = value
	case 0x5123:
		s.ChrBank4 = value
	case 0x5124:
		s.ChrBank5 = value
	case 0x5125:
		s.ChrBank6 = value
	case 0x5126:
		s.ChrBank7 = value
	case 0x5127:
		s.ChrBank8 = value
	case 0x5200:
		// vertical split (ignored)
	case 0x5203:
		s.IrqScanline = value
	case 0x5204:
		s.IrqEnabled = (value & 0x80) != 0
	case 0x5100:
		s.PrgMode = value & 0x03
	case 0x5101:
		s.ChrMode = value & 0x03
	case 0x5205:
		s.MultA = value
		s.MultResult = uint16(s.MultA) * uint16(s.MultB)
	case 0x5206:
		s.MultB = value
		s.MultResult = uint16(s.MultA) * uint16(s.MultB)
	}
}

func mmc5ReadCHR(m *Mapper, addr uint16) uint8 {
	s := &m.MMC5
	if s.ChrSize == 0 {
		return 0
	}
	banks := [8]uint8{s.ChrBank1, s.ChrBank2, s.ChrBank3, s.ChrBank4, s.ChrBank5, s.ChrBank6, s.ChrBank7, s.ChrBank8}
	slot := (uint32(addr) >> 10) & 7
	offset := uint32(addr) & 0x03FF
	bank := uint32(banks[slot]) % (s.ChrSize / 1024)
	idx := bank*1024 + offset
	if idx >= s.ChrSize {
		return 0
	}
	return s.Chr[idx]
}

func mmc5WriteCHR(m *Mapper, addr uint16, value uint8) {
	s := &m.MMC5
	if !s.ChrIsRAM {
		return
	}
	if s.ChrSize == 0 {
		return
	}
	s.Chr[uint32(addr)%s.ChrSize] = value
}

func mmc5ClockIRQ(m *Mapper) {
	// MMC5 IRQ is scanline-based; simplified
}

func mmc5ResetScanlineCounter(m *Mapper) {
	s := &m.MMC5
	s.FrameCount = 0
	s.InFrame = true
}

func mmc5SaveState(m *Mapper, buf []byte) int {
	s := &m.MMC5
	total := 4 + int(s.PrgSize) + 4 + 1 + 1 + 1 + 1 + 1 + 3 + 8 + 1 + 1 + 1 + 2 + 2 + 1 + 1 + 1 + 1 + mmc5PrgBankSize
	if s.ChrIsRAM {
		total += int(s.ChrSize)
	}
	if buf == nil {
		return total
	}
	off := 0
	putU32(buf[off:], s.PrgSize); off += 4
	if s.PrgSize > 0 {
		copy(buf[off:], s.PrgRom); off += int(s.PrgSize)
	}
	putU32(buf[off:], s.ChrSize); off += 4
	buf[off] = boolByte(s.ChrIsRAM); off++
	buf[off] = boolByte(s.HasBattery); off++
	buf[off] = uint8(s.Mirroring); off++
	buf[off] = s.PrgMode; off++
	buf[off] = s.ChrMode; off++
	buf[off] = s.PrgBank1; off++
	buf[off] = s.PrgBank2; off++
	buf[off] = s.PrgBank3; off++
	buf[off] = s.ChrBank1; off++
	buf[off] = s.ChrBank2; off++
	buf[off] = s.ChrBank3; off++
	buf[off] = s.ChrBank4; off++
	buf[off] = s.ChrBank5; off++
	buf[off] = s.ChrBank6; off++
	buf[off] = s.ChrBank7; off++
	buf[off] = s.ChrBank8; off++
	buf[off] = s.MultA; off++
	buf[off] = s.MultB; off++
	putU16(buf[off:], s.MultResult); off += 2
	buf[off] = s.FillTile; off++
	buf[off] = s.FillAttr; off++
	buf[off] = s.IrqScanline; off++
	buf[off] = boolByte(s.IrqEnabled); off++
	buf[off] = boolByte(s.IrqPending); off++
	putU16(buf[off:], s.FrameCount); off += 2
	buf[off] = boolByte(s.InFrame); off++
	copy(buf[off:], s.PrgRAM[:]); off += mmc5PrgBankSize
	if s.ChrIsRAM && s.ChrSize > 0 {
		copy(buf[off:], s.Chr); off += int(s.ChrSize)
	}
	return off
}

func mmc5LoadState(m *Mapper, buf []byte) bool {
	s := &m.MMC5
	if len(buf) < 4 {
		return false
	}
	off := 0
	prgSize := getU32(buf[off:]); off += 4
	if prgSize != s.PrgSize {
		return false
	}
	if s.PrgSize > 0 {
		copy(s.PrgRom, buf[off:off+int(s.PrgSize)]); off += int(s.PrgSize)
	}
	if len(buf) < off+4 {
		return false
	}
	chrSize := getU32(buf[off:]); off += 4
	if chrSize != s.ChrSize {
		return false
	}
	s.ChrIsRAM = buf[off] != 0; off++
	s.HasBattery = buf[off] != 0; off++
	s.Mirroring = Mirroring(buf[off]); off++
	s.PrgMode = buf[off]; off++
	s.ChrMode = buf[off]; off++
	s.PrgBank1 = buf[off]; off++
	s.PrgBank2 = buf[off]; off++
	s.PrgBank3 = buf[off]; off++
	s.ChrBank1 = buf[off]; off++
	s.ChrBank2 = buf[off]; off++
	s.ChrBank3 = buf[off]; off++
	s.ChrBank4 = buf[off]; off++
	s.ChrBank5 = buf[off]; off++
	s.ChrBank6 = buf[off]; off++
	s.ChrBank7 = buf[off]; off++
	s.ChrBank8 = buf[off]; off++
	s.MultA = buf[off]; off++
	s.MultB = buf[off]; off++
	s.MultResult = getU16(buf[off:]); off += 2
	s.FillTile = buf[off]; off++
	s.FillAttr = buf[off]; off++
	s.IrqScanline = buf[off]; off++
	s.IrqEnabled = buf[off] != 0; off++
	s.IrqPending = buf[off] != 0; off++
	s.FrameCount = getU16(buf[off:]); off += 2
	s.InFrame = buf[off] != 0; off++
	if len(buf) < off+mmc5PrgBankSize {
		return false
	}
	copy(s.PrgRAM[:], buf[off:off+mmc5PrgBankSize]); off += mmc5PrgBankSize
	if s.ChrIsRAM && s.ChrSize > 0 {
		if len(buf) < off+int(s.ChrSize) {
			return false
		}
		copy(s.Chr, buf[off:off+int(s.ChrSize)]); off += int(s.ChrSize)
	}
	return true
}

File: cores/java/src/main/java/nes/core/mappers/Mmc5.java
package nes.core.mappers;

import nes.core.Mapper;
import nes.core.Mirroring;

// Mapper 5 (MMC5). Port of cores/csharp/src/Mappers/Mmc5.cs.
public final class Mmc5 extends Mapper {
    private static final int PRG_BANK_SIZE = 8192;
    private static final int CHR_1K_SIZE = 1024;
    private static final int PRG_RAM_SIZE = 65536;
    private static final int PRG_RAM_WINDOW = 8192;

    private final byte[] _prgRom;
    private final int _prgSize;
    private byte[] _chr;
    private int _chrSize;
    private boolean _chrIsRam;
    private final byte[] _prgRam = new byte[PRG_RAM_SIZE];
    private final boolean _hasBattery;
    private byte _prgMode = 3;
    private byte _chrMode = 3;
    private byte _prgRamProtect1;
    private byte _prgRamProtect2;
    private byte _ntMirroring;
    private byte _fillTile;
    private byte _fillAttr;
    private byte _prgRamBank;
    private final byte[] _prgBanks = new byte[4];
    private final byte[] _chrBanks = new byte[8];
    private final byte[] _chrBanksEx = new byte[4];
    private byte _chrUpper;
    private byte _splitControl;
    private byte _splitYScroll;
    private byte _splitBank;
    private byte _irqScanline;
    private byte _irqControl;
    private byte _scanlineCounter;
    private boolean _irqPending;
    private boolean _inVblank;
    private byte _multA;
    private byte _multB;

    public Mmc5(byte[] prg, int prgSize, byte[] chr, int chrSize, int mirroring, boolean hasBattery) {
        mapperNum = 5;
        _prgSize = prgSize;
        _prgRom = new byte[prgSize > 0 ? prgSize : 1];
        if (prgSize > 0) System.arraycopy(prg, 0, _prgRom, 0, prgSize);
        if (chrSize > 0) {
            _chrSize = chrSize;
            _chr = new byte[chrSize];
            System.arraycopy(chr, 0, _chr, 0, chrSize);
            _chrIsRam = false;
        } else {
            _chrSize = 8192;
            _chr = new byte[8192];
            _chrIsRam = true;
        }
        _hasBattery = hasBattery;
        switch (mirroring) {
            case Mirroring.HORIZONTAL: _ntMirroring = 0x44; break;
            case Mirroring.VERTICAL: _ntMirroring = 0x50; break;
            case Mirroring.FOUR_SCREEN: _ntMirroring = (byte) 0xE4; break;
            case Mirroring.SINGLE_SCREEN0: _ntMirroring = 0x00; break;
            case Mirroring.SINGLE_SCREEN1: _ntMirroring = 0x55; break;
            case Mirroring.SINGLE_SCREEN2: _ntMirroring = (byte) 0xAA; break;
            case Mirroring.SINGLE_SCREEN3: _ntMirroring = (byte) 0xFF; break;
            default: _ntMirroring = 0x44; break;
        }
    }

    private int prgBankCount() { return _prgSize / PRG_BANK_SIZE > 0 ? _prgSize / PRG_BANK_SIZE : 1; }
    private int chr1kCount() { return _chrSize / CHR_1K_SIZE > 0 ? _chrSize / CHR_1K_SIZE : 1; }

    private int prgReadBank(int bank, int offset) {
        int count = prgBankCount();
        bank %= count;
        int idx = bank * PRG_BANK_SIZE + offset;
        if (idx >= _prgSize) return 0x00;
        return _prgRom[idx] & 0xFF;
    }

    private int product() { return (_multA & 0xFF) * (_multB & 0xFF); }

    private boolean prgRamWritesAllowed() { return _prgRamProtect1 == 0x02 && _prgRamProtect2 == 0x01; }

    private int decodeMirroring() {
        int s0 = _ntMirroring & 0x03;
        int s1 = (_ntMirroring >> 2) & 0x03;
        int s2 = (_ntMirroring >> 4) & 0x03;
        int s3 = (_ntMirroring >> 6) & 0x03;
        if (s0 == s1 && s1 == s2 && s2 == s3) {
            switch (s0) {
                case 0: return Mirroring.SINGLE_SCREEN0;
                case 1: return Mirroring.SINGLE_SCREEN1;
                case 2: return Mirroring.SINGLE_SCREEN2;
                default: return Mirroring.SINGLE_SCREEN3;
            }
        }
        if (s0 == 0 && s1 == 1 && s2 == 0 && s3 == 1) return Mirroring.HORIZONTAL;
        if (s0 == 0 && s1 == 0 && s2 == 1 && s3 == 1) return Mirroring.VERTICAL;
        if (s0 == 0 && s1 == 1 && s2 == 2 && s3 == 3) return Mirroring.FOUR_SCREEN;
        return Mirroring.HORIZONTAL;
    }

    @Override
    public int readPrg(int addr) {
        if (addr == 0x5205) return product() & 0xFF;
        if (addr == 0x5206) return (product() >> 8) & 0xFF;
        if (addr == 0x5204) {
            int status = _irqControl & 0x80;
            if (_inVblank) status |= 0x40;
            return status;
        }
        if (addr >= 0x6000 && addr < 0x8000) {
            int bank = (_prgRamBank & 0xFF) % (PRG_RAM_SIZE / PRG_RAM_WINDOW);
            int idx = bank * PRG_RAM_WINDOW + ((addr - 0x6000) & (PRG_RAM_WINDOW - 1));
            if (idx >= PRG_RAM_SIZE) return 0x00;
            return _prgRam[idx] & 0xFF;
        }
        int local = addr - 0x8000;
        int slot = local / PRG_BANK_SIZE;
        int offset = local & (PRG_BANK_SIZE - 1);
        if (slot == 3) {
            int last = prgBankCount() - 1;
            return prgReadBank(last, offset);
        }
        int reg = _prgBanks[slot] & 0xFF;
        if ((reg & 0x80) != 0) {
            int ramBank = (reg & 0x7F) % (PRG_RAM_SIZE / PRG_BANK_SIZE);
            int idx = ramBank * PRG_BANK_SIZE + offset;
            if (idx >= PRG_RAM_SIZE) return 0x00;
            return _prgRam[idx] & 0xFF;
        }
        int count = prgBankCount();
        int bank2 = reg % count;
        return prgReadBank(bank2, offset);
    }

    @Override
    public void writePrg(int addr, int value) {
        int v = value & 0xFF;
        if (addr == 0x5205) { _multA = (byte) v; return; }
        if (addr == 0x5206) { _multB = (byte) v; return; }
        if (addr == 0x5204) {
            _irqControl = (byte) (v & 0x80);
            if ((v & 0x80) == 0) _irqPending = false;
            return;
        }
        if (addr == 0x5203) { _irqScanline = (byte) v; return; }
        if (addr == 0x5200) { _splitControl = (byte) v; return; }
        if (addr == 0x5201) { _splitYScroll = (byte) v; return; }
        if (addr == 0x5202) { _splitBank = (byte) v; return; }
        switch (addr) {
            case 0x5100: _prgMode = (byte) (v & 0x03); return;
            case 0x5101: _chrMode = (byte) (v & 0x03); return;
            case 0x5102: _prgRamProtect1 = (byte) (v & 0x03); return;
            case 0x5103: _prgRamProtect2 = (byte) (v & 0x03); return;
            case 0x5104: return;
            case 0x5105: _ntMirroring = (byte) v; return;
            case 0x5106: _fillTile = (byte) v; return;
            case 0x5107: _fillAttr = (byte) v; return;
            case 0x5113: _prgRamBank = (byte) v; return;
            case 0x5114: _prgBanks[0] = (byte) v; return;
            case 0x5115: _prgBanks[1] = (byte) v; return;
            case 0x5116: _prgBanks[2] = (byte) v; return;
            case 0x5117: _prgBanks[3] = (byte) v; return;
            case 0x5120: _chrBanks[0] = (byte) v; return;
            case 0x5121: _chrBanks[1] = (byte) v; return;
            case 0x5122: _chrBanks[2] = (byte) v; return;
            case 0x5123: _chrBanks[3] = (byte) v; return;
            case 0x5124: _chrBanks[4] = (byte) v; return;
            case 0x5125: _chrBanks[5] = (byte) v; return;
            case 0x5126: _chrBanks[6] = (byte) v; return;
            case 0x5127: _chrBanks[7] = (byte) v; return;
            case 0x5128: _chrBanksEx[0] = (byte) v; return;
            case 0x5129: _chrBanksEx[1] = (byte) v; return;
            case 0x512A: _chrBanksEx[2] = (byte) v; return;
            case 0x512B: _chrBanksEx[3] = (byte) v; return;
            case 0x5130: _chrUpper = (byte) (v & 0x01); return;
        }
        if (addr >= 0x6000 && addr < 0x8000 && prgRamWritesAllowed()) {
            int bank = (_prgRamBank & 0xFF) % (PRG_RAM_SIZE / PRG_RAM_WINDOW);
            int idx = bank * PRG_RAM_WINDOW + ((addr - 0x6000) & (PRG_RAM_WINDOW - 1));
            if (idx < PRG_RAM_SIZE) _prgRam[idx] = (byte) v;
            return;
        }
        if (addr >= 0x8000 && addr < 0xE000 && prgRamWritesAllowed()) {
            int local = addr - 0x8000;
            int slot = local / PRG_BANK_SIZE;
            int offset = local & (PRG_BANK_SIZE - 1);
            if (slot < 3) {
                int reg = _prgBanks[slot] & 0xFF;
                if ((reg & 0x80) != 0) {
                    int ramBank = (reg & 0x7F) % (PRG_RAM_SIZE / PRG_BANK_SIZE);
                    int idx = ramBank * PRG_BANK_SIZE + offset;
                    if (idx < PRG_RAM_SIZE) _prgRam[idx] = (byte) v;
                }
            }
        }
    }

    @Override
    public int readChr(int addr) {
        int slot = addr / CHR_1K_SIZE;
        int offset = addr & (CHR_1K_SIZE - 1);
        int count = chr1kCount();
        int bankReg = _chrBanks[slot] & 0xFF;
        int bank = (((_chrUpper & 0xFF) << 8) | bankReg) % count;
        int idx = bank * CHR_1K_SIZE + offset;
        if (idx >= _chrSize) return 0x00;
        return _chr[idx] & 0xFF;
    }

    @Override
    public void writeChr(int addr, int value) {
        if (!_chrIsRam) return;
        int slot = addr / CHR_1K_SIZE;
        int offset = addr & (CHR_1K_SIZE - 1);
        int count = chr1kCount();
        int bankReg = _chrBanks[slot] & 0xFF;
        int bank = (((_chrUpper & 0xFF) << 8) | bankReg) % count;
        int idx = bank * CHR_1K_SIZE + offset;
        if (idx < _chrSize) _chr[idx] = (byte) value;
    }

    @Override
    public int mirrorMode() { return decodeMirroring(); }
    @Override
    public boolean chrIsRam() { return _chrIsRam; }
    @Override
    public boolean hasBattery() { return _hasBattery; }
    @Override
    public boolean irqPending() { return _irqPending; }

    @Override
    public void clockIrq() {
        _scanlineCounter++;
        if (_scanlineCounter >= 240) _inVblank = true;
        if (_scanlineCounter == (_irqScanline & 0xFF) && (_irqControl & 0x80) != 0) _irqPending = true;
    }

    @Override
    public void resetScanlineCounter() {
        _scanlineCounter = 0;
        _inVblank = false;
    }

    @Override
    public int saveState(byte[] buf) {
        int scalarSize = 19;
        int total = scalarSize + 4 + 8 + 4 + PRG_RAM_SIZE + (_chrIsRam ? _chrSize : 0);
        if (buf == null) return total;
        int p = 0;
        buf[p++] = _prgMode;
        buf[p++] = _chrMode;
        buf[p++] = _prgRamProtect1;
        buf[p++] = _prgRamProtect2;
        buf[p++] = _ntMirroring;
        buf[p++] = _fillTile;
        buf[p++] = _fillAttr;
        buf[p++] = _prgRamBank;
        buf[p++] = _chrUpper;
        buf[p++] = _splitControl;
        buf[p++] = _splitYScroll;
        buf[p++] = _splitBank;
        buf[p++] = _irqScanline;
        buf[p++] = _irqControl;
        buf[p++] = _scanlineCounter;
        buf[p++] = (byte) (_inVblank ? 1 : 0);
        buf[p++] = (byte) (_irqPending ? 1 : 0);
        buf[p++] = _multA;
        buf[p++] = _multB;
        for (int i = 0; i < 4; ++i) buf[p++] = _prgBanks[i];
        for (int i = 0; i < 8; ++i) buf[p++] = _chrBanks[i];
        for (int i = 0; i < 4; ++i) buf[p++] = _chrBanksEx[i];
        System.arraycopy(_prgRam, 0, buf, p, PRG_RAM_SIZE); p += PRG_RAM_SIZE;
        if (_chrIsRam && _chrSize > 0) {
            System.arraycopy(_chr, 0, buf, p, _chrSize);
            p += _chrSize;
        }
        return p;
    }

    @Override
    public boolean loadState(byte[] buf, int len) {
        int need = 19 + 4 + 8 + 4 + PRG_RAM_SIZE + (_chrIsRam ? _chrSize : 0);
        if (len < need) return false;
        int p = 0;
        _prgMode = buf[p++];
        _chrMode = buf[p++];
        _prgRamProtect1 = buf[p++];
        _prgRamProtect2 = buf[p++];
        _ntMirroring = buf[p++];
        _fillTile = buf[p++];
        _fillAttr = buf[p++];
        _prgRamBank = buf[p++];
        _chrUpper = buf[p++];
        _splitControl = buf[p++];
        _splitYScroll = buf[p++];
        _splitBank = buf[p++];
        _irqScanline = buf[p++];
        _irqControl = buf[p++];
        _scanlineCounter = buf[p++];
        _inVblank = buf[p++] != 0;
        _irqPending = buf[p++] != 0;
        _multA = buf[p++];
        _multB = buf[p++];
        for (int i = 0; i < 4; ++i) _prgBanks[i] = buf[p++];
        for (int i = 0; i < 8; ++i) _chrBanks[i] = buf[p++];
        for (int i = 0; i < 4; ++i) _chrBanksEx[i] = buf[p++];
        System.arraycopy(buf, p, _prgRam, 0, PRG_RAM_SIZE); p += PRG_RAM_SIZE;
        if (_chrIsRam && _chrSize > 0) {
            System.arraycopy(buf, p, _chr, 0, _chrSize);
            p += _chrSize;
        }
        return true;
    }
}
